core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -482,21 +482,28 @@ private[spark] class TaskSchedulerImpl(
     }
   }
 
-  /** Remove an executor from all our data structures and mark it as lost */
+  /**
+   * Remove an executor from all our data structures and mark it as lost. If the executor's loss
+   * reason is not yet known, do not yet remove its association with its host.
+   */
   private def removeExecutor(executorId: String, reason: ExecutorLossReason) {
     activeExecutorIds -= executorId
     val host = executorIdToHost(executorId)
     val execs = executorsByHost.getOrElse(host, new HashSet)
-    execs -= executorId
-    if (execs.isEmpty) {
-      executorsByHost -= host
-      for (rack <- getRackForHost(host); hosts <- hostsByRack.get(rack)) {
-        hosts -= host
-        if (hosts.isEmpty) {
-          hostsByRack -= rack
+
+    if (reason != LossReasonPending) {
```

Comment:
But executorIdToHost is updated outside of your check, on line 507, so it seems like even with this change host will be null? But this question is mostly irrelevant given my other question about why not just do this the simpler way I proposed?